repoconfig library to read git directories and .labels files

foxish opened the pull request:
This addresses https://github.com/signadot/signadot/issues/5546. Adds a package that helps read git config and labels from a repo. It is hooked up to `signadot test run` for now (when run with no args). Example run below.

```
go run ./cmd/signadot/main.go test run
Reading config from: /Users/anirudhramanathan/Documents/cli/.signadot/config.yaml
Found 4 test files:
  tests/fixtures/smart-tests-git/backend/team1/file1.star
    Labels:
      suite=integration
      area=backend
      team=team1
      signadot.com/run-id=skthedwn
  tests/fixtures/smart-tests-git/backend/team2/file2.star
    Labels:
      area=backend
      team=team2
      signadot.com/run-id=skthedwn
      suite=integration
  tests/fixtures/smart-tests-git/backend/test.star
    Labels:
      suite=integration
      area=backend
      signadot.com/run-id=skthedwn
  tests/fixtures/smart-tests-git/frontend/file3.star
    Labels:
      suite=integration
      team=frontend
      signadot.com/run-id=skthedwn
```

scott-cotton commented on go.mod:
```diff
@@ -1,10 +1,11 @@
 module github.com/signadot/cli
 
-go 1.23
+go 1.23.0
```
If you keep this as 1.23 then it allows us to upgrade minor versions of Go more easily.

foxish replied:
go mod tidy is doing this looks like - is there a way you're aware of that prevents this?

scott-cotton replied:
arg they changed it so it overwrites that now by default.

I'm ok with removing this discussion from this PR, but I think we need to either edit the go.mod manually after go mod tidy or explicitly track minor version upgrades.

scott-cotton commented on internal/repoconfig/labels.go:
```diff
@@ -0,0 +1,101 @@
+			continue
+		}
+
+		parts := strings.SplitN(line, "=", 2)
```
would suggest https://pkg.go.dev/strings#Cut

foxish replied:
done, thanks!

scott-cotton commented on internal/config/test.go:
```diff
@@ -45,8 +45,7 @@ type TestRun struct {
 func (cfg *TestRun) AddFlags(cmd *cobra.Command) {
-	cmd.Flags().StringVarP(&cfg.Cluster, "cluster", "c", "", "cluster name (required)")
-	cmd.MarkFlagRequired("cluster")
+	cmd.Flags().StringVarP(&cfg.Cluster, "cluster", "c", "", "cluster name (required for test execution)")
```
Looks like this will break e2es.

scott-cotton replied:
never mind, didn't read it closely enough

scott-cotton commented on internal/repoconfig/config.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	configPath := filepath.Join(repoRoot, ".signadot", "config.yaml")
+	fmt.Printf("Reading config from: %s\n", configPath)
```
debug print

scott-cotton commented on internal/repoconfig/config.go:
```diff
@@ -0,0 +1,187 @@
+	testFileMap := make(map[string]TestFile)
+
+	// Sort paths by length to process parent directories first
+	sortedPaths := sortPathsByLength(cfg.SmartTests)
```
Probably best to clean the paths from the config file to avoid noise like ../dir/.. and ././ etc

scott-cotton commented on internal/repoconfig/config.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// FindTestFiles finds all test files in the tests directories
+func FindTestFiles(startPath string, cfg *Config) ([]TestFile, error) {
```
This function is long and has many nested levels of loops/funcs. Can we break it down some more?

foxish replied:
Modularized it a bit - hopefully it's more maintainable now.

foxish commented:
Fixed the CR comments and tried to break things down further. Also added some support for directly specifying the file path in config (as opposed to directory) and corresponding test. PTAL

daniel-de-vera commented:
These changes LGTM and will work on top of this.
@scott-cotton, given you were the initial reviewer, I wait for you to add the rubber stamp.
